### PR TITLE
[Locations] Finish backend/frontend adaptation for new location values

### PR DESCRIPTION
**Pull request by jshoe**
### Description
- Update Metadata components to display the location objects (generalized for if we have any other JSON metadata types in the future)
- Add GeoSearchInputBox to MetadataInput
- Modify GeoSearchInputBox to work with showing a value from parent props
- Finish check_and_set_location_type validation to associate with Locations table
- Add "data migration" for adding a "new_collection_location" MetadataField. Adds the field to the project "CI" for admins (can expand once we start releasing to users).
- Later: Updating component on the /samples/upload page to pop out of the inner modal t_t

### Demo
May-21-2019 11-09-44
May-21-2019 11-10-23
May-21-2019 11-14-35
May-21-2019 11-18-44
May-21-2019 11-19-31
May-21-2019 11-58-52

### Test and Reproduce
- Pull the branch, run migration, go to project CI, test out the flows above. You can verify the values in the console like Sample.find(10).metadata.last.location_id.

### CODE REVIEW

File: app/assets/src/components/common/DetailsSidebar/SampleDetailsMode/MetadataTab.jsx [hunk 1]
```diff
@@ -119,7 +119,10 @@ class MetadataTab extends React.Component {
     return val === undefined || val === null || val === "" ? (
       <div className={cs.emptyValue}>--</div>
     ) : (
-      <div className={cs.metadataValue}>{val}</div>
+      <div className={cs.metadataValue}>
+        {/* If asked to display an arbitrary object, try the name field */}
+        {isObject(val) && val.name ? val.name : val}
```
Comment (tfrcarvalho) [inline app/assets/src/components/common/DetailsSidebar/SampleDetailsMode/MetadataTab.jsx]: Does it still make sense to display the object? In what circumstances would that happen?

Comment (jshoe) [reply app/assets/src/components/common/DetailsSidebar/SampleDetailsMode/MetadataTab.jsx]: probably better not to try to display the object (you'll get `Objects are not valid as a React child`). I'll change to `isObject(val) ? val.name : val`

File: db/migrate/20190521175543_add_new_collection_location.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+class AddNewCollectionLocation < ActiveRecord::Migration[5.1]
+  # Data migration to add new intelligent collection location MetadataField
+
+  def up
+    existing = MetadataField.find_by(name: "new_collection_location")
+    unless existing
+      field = MetadataField.create(name: "new_collection_location", display_name: "New Collection Location", description: "Location sample was originally collected", base_type: 3, group: "Sample", host_genomes: HostGenome.all)
```
Comment (tfrcarvalho) [inline db/migrate/20190521175543_add_new_collection_location.rb]: What is the migraiton plan for the transition from collection_location to new_collection_location?

File: app/helpers/error_helper.rb [hunk 3]
```diff
@@ -31,11 +31,16 @@ def self.missing_required_metadata(sample, missing_metadata_fields)
     end
   end
 
+  LOCATION_INVALID_ERROR = "Please input a location.".freeze
   DATE_INVALID_ERROR_HUMAN = "Please input a date in the format YYYY-MM or MM/YYYY. (for human samples)".freeze
   DATE_INVALID_ERROR = "Please input a date in the format YYYY-MM-DD, YYYY-MM, MM/DD/YY, or MM/YYYY.".freeze
   NUMBER_INVALID_ERROR = "Please input a number.".freeze
 
   def get_field_error(field, is_human = false)
+    if field.base_type == 3
```
Comment (tfrcarvalho) [inline app/helpers/error_helper.rb]: There are not constants to use instead of this magic numbers?

Comment (jshoe) [reply app/helpers/error_helper.rb]: fixed 👍

File: db/migrate/20190521175543_add_new_collection_location.rb (hunk 2 above)
Comment (jshoe) [reply db/migrate/20190521175543_add_new_collection_location.rb]: added "Data Migration Plan" here https://czi.quip.com/gzWdAafIX8Yb/LocationMapping-Sprint-Plan

**Pull request merged**